### PR TITLE
Elg 263

### PR DESCRIPTION
**Pull request by gcleary**
This change depends on the following PR in embedded-protocol: https://github.com/SkyhookWireless/embedded-protocol/pull/6

So far this has Used_aps bitmap in protocol, and uses the info during filtering.

### CODE REVIEW

**Comment by gcleary** [comment]
Updated the scoring code to use the Used_aps info. 90% of the change is in find_best_match()

**Comment by tboinske-skyhook** [comment]
Do you intend to create a PR for the associated `embedded-protocol` change? Assuming so, you should include a link to it in the description of this PR, since it represents a dependency.

**Comment by gcleary** [comment]
> Do you intend to create a PR for the associated `embedded-protocol` change? Assuming so, you should include a link to it in the description of this PR, since it represents a dependency.



> Do you intend to create a PR for the associated `embedded-protocol` change? Assuming so, you should include a link to it in the description of this PR, since it represents a dependency.

Yes, done.

**Comment by gcleary** [comment]
> Line #147 (which I realize is not changed in this PR):
> 
> ```c
> /*! \brief try to reduce AP by filtering out based on diversity of rssi
> ```
> 
> What does it mean to "reduce AP"? I assume you mean "reduce the AP count". By how much? I.e., how much are we trying to reduce it? Is the set of APs within the context assumed to be sorted in order of RSSI? (It appears so.) I.e., this function-level comment needs to be fleshed out. It's hard to understand the code details without having a good summary of the overall strategy.
> 
> Cache filtering and matching has grown to be quite complex. It's good that there's unit test code, but how complete is it? I would normally expect to see calls to `assert()` (or something similar) in the unit test code, but there are none. I would have more confidence if the unit test code included lots of well defined, hand-crafted test-cases, including corner cases, with calls to assert for each one in order to verify that the results are as expected. Presently I don't see this sort of organization in the unit-test code. If it's actually there and I'm just overlooking it, please let me know. Otherwise, I think we need to bolster that piece.

I will address Unit Tests under ticket ELG-271. Improved function level descriptions are addressed in PR ELG-263.

File: libel/protocol/proto.c [hunk 1]
```diff
@@ -523,13 +533,24 @@ int32_t deserialize_response(Sky_ctx_t *ctx, uint8_t *buf, uint32_t buf_len, Sky
...
+        loc->hpe = (uint16_t)rs.hpe;
+        loc->location_source = (Sky_loc_source_t)rs.source;
+        // Extract Used info for each AP from the Used_aps bytes
+        for (a = 0; a < ctx->ap_len; a++) {
+            i = rs.used_aps.size - 1 - (a / 8);
+            // LOGFMT(ctx, SKY_LOG_LEVEL_DEBUG, "Used: AP %d, Byte: 0x%02X idx: %d:%d", a, rs.used_aps.bytes[i], i, (a % 8))
+            if (a < rs.used_aps.size * 8)
+                ctx->beacon[a].ap.property.used = rs.used_aps.bytes[i] & (1 << (a % 8)) ? 1 : 0;
+            else
+                ctx->beacon[a].ap.property.used = 0;
```
Comment (tboinske-skyhook) [inline libel/protocol/proto.c:551]: This case is not expected to happen by design, but you're guarding against it just to be defensive, right? If so, I suggest adding a comment to this effect.

File: libel/workspace.h [hunk 2]
```diff
@@ -80,6 +82,7 @@ typedef struct sky_ctx {
     int16_t connected; /* which beacon is conneted (-1 == none) */
     Gps_t gps; /* GNSS info */
     /* Assume worst case is that beacons and gps info takes twice the bare structure size */
+    int16_t bestput; /* cacheline with best match for saving */
```
Comment (tboinske-skyhook) [inline libel/workspace.h:85]: What's the rationale for the name "bestput"? Seems a little funny, but perhaps I'm just not getting it.

File: libel/protocol/proto.c (hunk 1 above)
Comment (gcleary) [reply libel/protocol/proto.c:551]: It could happen if we add an optimization on the elg server side. When the bitvector is converted to bytes in the protocol, we could drop leading zero bytes. Perhaps protobuf already does this? Anyway, this conditional makes sure that the used boolean value in the ap structure is set to false if the zero bytes are not actually send on the wire.

I still agree that a comment is a good.

File: libel/workspace.h (hunk 2 above)
Comment (gcleary) [reply libel/workspace.h:85]: In the 1.1.4 release, find_best_match was called twice. Once to look for a match before a server request was made, and again after a server response is received to look for the best place to save this new location.

In this version, I call find_best_match only once before making a request, but in anticipation of needing to know where to save a response, this code saves that best cacheline for the save in the workspace. Only one search is done, but two results are found, a cacheline to retrieve from, and a cacheline to potentially save to. bestput means "best line for save".

**Comment by gcleary** [comment]
Obsolete branch. Addressing the Used AP work in EC-63.
